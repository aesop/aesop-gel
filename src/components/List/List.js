import React, { forwardRef } from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import styles from './List.module.css';

const List = forwardRef(
  ({ className, items, listItemClassName, theme }, ref) => {
    const classSet = cx(styles.base, styles[theme], className);

    console.log('ref', ref, items);

    return (
      <ul className={classSet} ref={ref}>
        {items.map(({ content, id }) => (
          <li className={cx(styles.item, listItemClassName)} key={id}>
            {content}
          </li>
        ))}
      </ul>
    );
  },
);

List.propTypes = {
  className: PropTypes.string,
  items: PropTypes.arrayOf(
    PropTypes.shape({
      content: PropTypes.oneOfType([PropTypes.node, PropTypes.element])
        .isRequired,
      id: PropTypes.string.isRequired,
    }),
  ).isRequired,
  listItemClassName: PropTypes.string,
  theme: PropTypes.oneOf(['dark', 'light']),
};

List.defaultProps = {
  className: undefined,
  items: undefined,
  listItemClassName: undefined,
  theme: 'dark',
};

export default List;
