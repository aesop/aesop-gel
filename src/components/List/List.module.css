@import 'styles/variables/index';

.base {
  padding-left: 0;
  margin-top: 0;
  margin-bottom: 0;
  list-style: none;
}

.item {
  &:not(:last-child) {
    margin-bottom: var(--layout-sm-spacing);
  }
}
