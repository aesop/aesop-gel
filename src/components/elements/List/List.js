import React from 'react';
import cx from 'classnames';
import uuidv4 from 'uuid/v4';
import Props from './List.types';
import styles from './List.module.css';

const List: React.FunctionComponent<Props> = ({
  className,
  items,
  listItemclassName,
}) => {
  const classSet: string = cx(styles.base, className);

  return (
    <ul className={classSet}>
      {items.map(({ content, id = uuidv4() }) => (
        <li className={cx(styles.item, listItemClassSet)} key={id}>
          {content}
        </li>
      ))}
    </ul>
  );
};

export default List;
