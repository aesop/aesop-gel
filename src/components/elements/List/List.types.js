import { ClassSetType } from '~/types';

interface ItemType {
  content: string | React.ReactNode;
  id?: string | undefined;
}

export interface Props extends ClassSetType {
  items: ItemType[];
  listItemClassSet?: string | undefined;
}

export { Props as default }; // eslint-disable-line no-undef
