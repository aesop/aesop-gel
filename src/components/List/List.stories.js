import React from 'react';
import { storiesOf } from '@storybook/react';
import { object, select } from '@storybook/addon-knobs';
import List from './List';

storiesOf('Components/List', module).add('Base component', () => (
  <List
    items={object('items', [
      { content: 'Fragrance', id: 'fragrance' },
      { content: 'Gifts', id: 'gifts' },
    ])}
    theme={select('theme', ['dark', 'light'], 'dark')}
  />
));
