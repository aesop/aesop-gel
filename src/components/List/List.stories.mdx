import { Meta, Story, Preview, Props } from '@storybook/addon-docs/blocks';
import { withKnobs } from '@storybook/addon-knobs';
import * as knobs from '@storybook/addon-knobs';
import List from './List';
import ListFixture from './List.fixture';

<Meta
  title="Components/List"
  component={List}
  decorators={[withKnobs]}
/>

# List

<Preview>
  <Story name="Base component">
    <List
      items={knobs.object('items', [
        { content: 'Fragrance', id: 'fragrance' },
        { content: 'Gifts', id: 'gifts' },
      ])}
      theme={knobs.select('theme', ['dark', 'light'], 'dark')}
    />
  </Story>
</Preview>

## API

<Props of={List} />
