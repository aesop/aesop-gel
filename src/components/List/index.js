/* istanbul ignore file */
import { List } from './List';

export { List };
