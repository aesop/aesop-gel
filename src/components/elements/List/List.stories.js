import React from 'react';
import { storiesOf } from '@storybook/react';
import { object } from '@storybook/addon-knobs';
import List from './List';

storiesOf('Elements.List', module).add('Base component', () => (
  <List
    items={object('items', [
      { content: 'Fragrance', id: 'fragrance' },
      { content: 'Gifts', id: 'gifts' },
    ])}
  />
));
