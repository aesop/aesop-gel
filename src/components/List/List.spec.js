import React from 'react';
import renderer from 'react-test-renderer';
import { select } from '@storybook/addon-knobs';
import List from './List';

describe('Component - Element - List', () => {
  it('should be defined', () => {
    expect(List).toBeDefined();
  });

  it('renders base component correctly', () => {
    const tree = renderer
      .create(
        <List
          items={[
            { content: 'Fragrance', id: 'fragrance' },
            { content: 'Gifts', id: 'gifts' },
          ]}
          theme={select('theme', ['dark', 'light'], 'dark')}
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
