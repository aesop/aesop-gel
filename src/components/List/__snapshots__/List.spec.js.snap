// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component - Element - List should render base component correctly 1`] = `
<div>
  <ul
    class="base dark"
  >
    <li
      class="item"
    >
      Fragrance
    </li>
    <li
      class="item"
    >
      Gifts
    </li>
  </ul>
</div>
`;
